MediaGalleryMetadata/Model/Gif/ExtractMetadata.php

```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryMetadata\Model\Gif;
+
+use Magento\MediaGalleryMetadataApi\Api\ExtractMetadataInterface;
+use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterface;
+use Magento\MediaGalleryMetadataApi\Api\Data\MetadataInterfaceFactory;
+use Magento\MediaGalleryMetadataApi\Model\MetadataReaderInterface;
+
+/**
+ * Extract metadata from the asset by path
+ */
+class ExtractMetadata implements ExtractMetadataInterface
+{
+    /**
+     * @var FileReader
+     */
+    private $fileReader;
+
+    /**
+     * @var MetadataReaderInterface[]
+     */
+    private $readers;
+
+    /**
+     * @var MetadataInterfaceFactory
+     */
+    private $metadataFactory;
+
+    /**
+     * @param FileReader $fileReader
+     * @param MetadataInterfaceFactory $metadataFactory
+     * @param MetadataReaderInterface[] $readers
+     */
+    public function __construct(
+        FileReader $fileReader,
+        MetadataInterfaceFactory $metadataFactory,
+        array $readers
+    ) {
+        $this->readers = $readers;
+        $this->fileReader = $fileReader;
+        $this->metadataFactory = $metadataFactory;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(string $path): MetadataInterface
+    {
+        $title = '';
+        $description = '';
+        $keywords = [];
+        if (!$this->fileReader->isApplicable($path)) {
+            return $this->metadataFactory->create([
+                'title' => $title,
+                'description' => $description,
+                'keywords' => $keywords
+            ]);
+        }
```

Comment:
Instead of returning an empty MetaDataInterface wouldn't be better to throw an exception and on your client, you just handle that and continue?

Reply:
I don't think there are any differences for us if the metadata fields are empty or metadata is missing. To have only one flow of processing this situation I decided to always return a MetadataInterface. The MetadataInterface contains the information that the reader was able to retrieve.

Also, this approach provides more flexibility to extensions using plugins